core/src/main/java/io/questdb/cairo/wal/seq/TableWriterPressureControlImpl.java

```diff
@@ -0,0 +1,154 @@
+
+        if (memoryPressureRegulationValue == Integer.MAX_VALUE) {
+            // already at max parallelism, can't go more optimistic
+            return false;
+        }
+        if (memoryPressureRegulationValue < 0) {
+            // was in backoff, go back to parallelism = 1
+            memoryPressureRegulationValue = 1;
+            return true;
+        }
+        if (MEM_PRESSURE_RND.nextInt(4) == 0) { // 25% chance to double parallelism
+            int beforeDoubling = memoryPressureRegulationValue;
+            memoryPressureRegulationValue *= PARTITION_COUNT_SCALE_UP_FACTOR;
+            if (memoryPressureRegulationValue < beforeDoubling) {
+                // overflow
+                memoryPressureRegulationValue = Integer.MAX_VALUE;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Applies anti-OOM measures if possible, either by reducing job parallelism, or applying backoff.<br>
+     * If it was possible to apply more measures, returns true → the operation can retry.<br>
+     * If all measures were exhausted, returns false → the operation should now fail.
+     */
+    @Override
+    public void onOutOfMemory() {
+        long inflightRows = inflightBlockRowCount;
+        inflightBlockRowCount = maxBlockRowCount = Math.max(1, inflightRows / TXN_COUNT_SCALE_DOWN_FACTOR);
+
+        if (maxRecordedInflightPartitions == 1 && inflightTxnCount <= 1) {
+            // There was no parallelism and no multi transaction block
+            if (memoryPressureRegulationValue <= -5) {
+                // Maximum backoff already tried => fail
+                walBackoffUntil = -1;
+                return;
+            }
+            if (memoryPressureRegulationValue > 0) {
+                // Switch from reducing parallelism to backoff
+                memoryPressureRegulationValue = -1;
+            } else {
+                // Increase backoff counter
+                memoryPressureRegulationValue--;
+            }
+            int delayMillis = MEM_PRESSURE_RND.nextInt(4_000);
+            walBackoffUntil = getTicks() + delayMillis;
+            return;
+        }
+        // There was some parallelism, halve max parallelism
+        walBackoffUntil = -1;
+        memoryPressureRegulationValue = maxRecordedInflightPartitions / PARTITION_COUNT_SCALE_DOWN_FACTOR;
+        maxRecordedInflightPartitions = 1;
+    }
+
+    @Override
+    public void updateInflightPartitions(int count) {
+        maxRecordedInflightPartitions = Math.max(maxRecordedInflightPartitions, count);
+    }
+
+    @Override
+    public void setMaxBlockRowCount(int count) {
+        maxBlockRowCount = count;
+    }
+
+    @Override
+    public void updateInflightTxnBlockSize(long txnCount, long rowCount) {
+        inflightBlockRowCount = rowCount;
+        inflightTxnCount = txnCount;
+    }
+
+    private static long getTicks() {
+        return MillisecondClockImpl.INSTANCE.getTicks();
```

Comment:
the clock could also come from config so we can have this logic  (whatever the logic is) tested